File: main/src/main/java/lavaplayer/source/youtube/YoutubeSearchProvider.java
package lavaplayer.source.youtube;

import lavaplayer.tools.DataFormatTools;
import lavaplayer.tools.ExceptionTools;
import lavaplayer.tools.JsonBrowser;
import lavaplayer.tools.ThumbnailTools;
import lavaplayer.tools.http.ExtendedHttpConfigurable;
import lavaplayer.tools.io.HttpClientTools;
import lavaplayer.tools.io.HttpInterface;
import lavaplayer.tools.io.HttpInterfaceManager;
import lavaplayer.track.*;
import org.apache.http.client.methods.CloseableHttpResponse;
import org.apache.http.client.methods.HttpPost;
import org.apache.http.entity.StringEntity;
import org.apache.http.util.EntityUtils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.function.Function;

import static java.nio.charset.StandardCharsets.UTF_8;

/**
 * Handles processing YouTube searches.
 */
public class YoutubeSearchProvider implements YoutubeSearchResultLoader {
    private static final Logger log = LoggerFactory.getLogger(YoutubeSearchProvider.class);

    private final HttpInterfaceManager httpInterfaceManager;

    public YoutubeSearchProvider() {
        this.httpInterfaceManager = HttpClientTools.createCookielessThreadLocalManager();
    }

    public ExtendedHttpConfigurable getHttpConfiguration() {
        return httpInterfaceManager;
    }

    /**
     * @param query Search query.
     * @return Playlist of the first page of results.
     */
    @Override
    public AudioItem loadSearchResult(String query, Function<AudioTrackInfo, AudioTrack> trackFactory) {
        log.debug("Performing a search with query {}", query);

        try (HttpInterface httpInterface = httpInterfaceManager.get()) {
            HttpPost post = new HttpPost(YoutubeConstants.SEARCH_URL);

            StringEntity payload = new StringEntity(String.format(YoutubeConstants.SEARCH_PAYLOAD, query.replace("\"", "\\\"")), "UTF-8");
            post.setEntity(payload);
            try (CloseableHttpResponse response = httpInterface.execute(post)) {
                HttpClientTools.assertSuccessWithContent(response, "search response");

                String responseText = EntityUtils.toString(response.getEntity(), UTF_8);

                JsonBrowser jsonBrowser = JsonBrowser.parse(responseText);
                return extractSearchResults(jsonBrowser, query, trackFactory);
            }
        } catch (Exception e) {
            throw ExceptionTools.wrapUnfriendlyExceptions(e);
        }
    }

    private AudioItem extractSearchResults(JsonBrowser jsonBrowser, String query,
                                           Function<AudioTrackInfo, AudioTrack> trackFactory) {
        List<AudioTrack> tracks;
        log.debug("Attempting to parse results from search page");
        try {
            tracks = extractSearchPage(jsonBrowser, trackFactory);
        } catch (IOException e) {
            throw new RuntimeException(e);
        }

        if (tracks.isEmpty()) {
            return AudioReference.NO_TRACK;
        } else {
            return new BasicAudioTrackCollection("Search results for: " + query, new AudioTrackCollectionType.SearchResult(query), tracks, null);
        }
    }

    private List<AudioTrack> extractSearchPage(JsonBrowser jsonBrowser, Function<AudioTrackInfo, AudioTrack> trackFactory) throws IOException {
        ArrayList<AudioTrack> list = new ArrayList<>();
        jsonBrowser.get("contents")
            .get("sectionListRenderer")
            .get("contents")
            .index(0)
            .get("itemSectionRenderer")
            .get("contents")
            .values()
            .forEach(jsonTrack -> {
                AudioTrack track = extractPolymerData(jsonTrack, trackFactory);
                if (track != null) {
                    list.add(track);
                }
            });
        return list;
    }

    private AudioTrack extractPolymerData(JsonBrowser json, Function<AudioTrackInfo, AudioTrack> trackFactory) {
        json = json.get("compactVideoRenderer");
        if (json.isNull()) {
            return null; // Ignore everything which is not a track
        }

        String title = json.get("title").get("runs").index(0).get("text").text();
        String author = json.get("longBylineText").get("runs").index(0).get("text").text();
        if (json.get("lengthText").isNull()) {
            return null; // Ignore if the video is a live stream
        }

        long duration = DataFormatTools.durationTextToMillis(json.get("lengthText").get("runs").index(0).get("text").text());
        String videoId = json.get("videoId").text();

        AudioTrackInfo info = new AudioTrackInfo(
            title,
            author,
            duration,
            videoId,
            false,
            YoutubeConstants.WATCH_URL_PREFIX + videoId,
            ThumbnailTools.extractYouTube(json, videoId)
        );

        return trackFactory.apply(info);
    }
}

File: main/src/main/java/lavaplayer/source/youtube/YoutubeSearchProvider.kt
package lavaplayer.source.youtube

import lavaplayer.tools.DataFormatTools.durationTextToMillis
import lavaplayer.tools.ExceptionTools
import lavaplayer.tools.ThumbnailTools
import lavaplayer.tools.http.ExtendedHttpConfigurable
import lavaplayer.tools.io.HttpClientTools
import lavaplayer.tools.io.HttpInterfaceManager
import lavaplayer.tools.json.JsonBrowser
import lavaplayer.tools.json.JsonBrowser.Companion.parse
import lavaplayer.track.*
import mu.KotlinLogging
import org.apache.http.client.methods.HttpPost
import org.apache.http.entity.StringEntity
import org.apache.http.util.EntityUtils
import java.io.IOException
import java.nio.charset.StandardCharsets

/**
 * Handles processing YouTube searches.
 */
class YoutubeSearchProvider : YoutubeSearchResultLoader {
    companion object {
        private val log = KotlinLogging.logger { }
    }

    private val httpInterfaceManager: HttpInterfaceManager = HttpClientTools.createCookielessThreadLocalManager()

    override val httpConfiguration: ExtendedHttpConfigurable
        get() = httpInterfaceManager

    /**
     * @param query Search query.
     * @return Playlist of the first page of results.
     */
    override fun loadSearchResult(query: String, trackFactory: AudioTrackFactory): AudioItem {
        log.debug { "Performing a search with query $query" }
        try {
            httpInterfaceManager.get().use { httpInterface ->
                val post = HttpPost(YoutubeConstants.SEARCH_URL)
                post.entity = StringEntity(YoutubeConstants.SEARCH_PAYLOAD.format(query.replace("\"", "\\\"")), "UTF-8")

                httpInterface.execute(post).use { response ->
                    HttpClientTools.assertSuccessWithContent(response, "search response")
                    val responseText = EntityUtils.toString(response.entity, StandardCharsets.UTF_8)
                    val jsonBrowser = parse(responseText)
                    return extractSearchResults(jsonBrowser, query, trackFactory)
                }
            }
        } catch (e: Exception) {
            throw ExceptionTools.wrapUnfriendlyException(e)
        }
    }

    private fun extractSearchResults(
        jsonBrowser: JsonBrowser,
        query: String,
        trackFactory: AudioTrackFactory
    ): AudioItem {
        log.debug { "Attempting to parse results from search page" }

        val tracks: MutableList<AudioTrack> = jsonBrowser
            .runCatching { extractSearchPage(this, trackFactory) }
            .onFailure { throw RuntimeException(it) }
            .getOrThrow()

        return if (tracks.isEmpty()) {
            AudioReference.NO_TRACK
        } else {
            BasicAudioTrackCollection(
                "Search results for: $query",
                AudioTrackCollectionType.SearchResult(query),
                tracks,
                null
            )
        }
    }

    @Throws(IOException::class)
    private fun extractSearchPage(
        jsonBrowser: JsonBrowser,
        trackFactory: AudioTrackFactory
    ): MutableList<AudioTrack> {
        return jsonBrowser["contents"]["sectionListRenderer"]["contents"][0]["itemSectionRenderer"]["contents"]
            .values()
            .mapNotNull { extractPolymerData(it, trackFactory) }
            .toMutableList()
    }

    private fun extractPolymerData(json: JsonBrowser, trackFactory: AudioTrackFactory): AudioTrack? {
        val renderer = json["compactVideoRenderer"].takeUnless { it.isNull }
            ?: return null // Ignore everything which is not a track

        /* Ignore if the video is a live stream */
        if (renderer["lengthText"].isNull) {
            return null
        }

        val videoId = renderer["videoId"].text

        /* create the audio track. */
        val info = AudioTrackInfo(
            title = renderer["title"]["runs"][0]["text"].safeText,
            author = renderer["longBylineText"]["runs"][0]["text"].safeText,
            length = durationTextToMillis(renderer["lengthText"]["runs"][0]["text"].safeText),
            identifier = videoId!!,
            uri = "${YoutubeConstants.WATCH_URL_PREFIX}$videoId",
            artworkUrl = ThumbnailTools.extractYouTube(renderer, videoId)
        )

        return trackFactory.create(info)
    }
}
